### DISCUSSION TITLE
HA event entities way forward

### DISCUSSION THREAD

**Discussion by Koenkk** [Koenkk/zigbee2mqtt#25061] [General]
Z2M 1.42.0 added support for the Home Assistant event entities. This was done in https://github.com/Koenkk/zigbee2mqtt/pull/24233. The current implementation has some limitations, it doesn't work for all actions atm. The current patterns are supported for now:

https://github.com/Koenkk/zigbee2mqtt/blob/5b907f2f77490cfcff1a39f5b731d508d0344663/lib/extension/homeassistant.ts#L39

My initial thought was to add a new `action_data` property for this (https://github.com/Koenkk/zigbee2mqtt/pull/24233#discussion_r1817842276), but this will add a lot of duplication, so not a good option.

Therefore I'm now thinking to move the patterns that are now in `homeassistant.ts` into the `e.action()` exposes.

What do you think? @mundschenk-at @Nerivec @Drafteed

**Comment by Nerivec** [comment]
Whatever way is picked, I think it should follow:
- The less duplication of values/props the better, it needs to be easy to maintain...
- Avoid splitting payloads into several props if the user is going to always recombine them in automation triggers anyway (pointless)
- The "final behavior" should resemble the behavior of regular action (zigbee2mqtt-frontend) enough that it avoids too much confusion on usage (too many patterns...)
- Should have a nil-to-limited footprint when HA extension is disabled (non-HA users)

**Reply by mundschenk-at** [reply]
I can see the reasoning, but I disagree. The frontend just follows the current MQTT topic and should not be normative. The HA event API works differently and should embrace it. Having the action itself (`event_type`) and attributes separate also allows for proper localization in the HA frontend.

**Reply by Sineos** [reply]
> Should have a nil-to-limited footprint when HA extension is disabled (non-HA users)

Without being able to contribute much to the technical side, I have a humble request:

Much of Zigbee2MQTT's beauty comes from the fact that it is agnostic to the home-automation ecosystem to which it is connected. With these ecosystems being "proprietary" to some degree, I kindly ask to not lose sight of your heritage as Zigbee2**MQTT**.

**Comment by mundschenk-at** [comment]
The current implementation should work on a basic level for almost all current actions, except for those like [511.344](https://github.com/Koenkk/zigbee-herdsman-converters/blob/fcd5cf77df3b0b07c039b840f3faa6c70d0dba04/src/devices/iluminize.ts#L186) which somehow adds endpoints to the actual payload string but not the `exposes` (I have to investigate this, just got notified a few hours ago), in that it passes the current action value through as the HA `event_type`. However, if we want to have nice even types and attributes in HA, the actions are currently parsed both on the Z2M side to define valid `event_type` values for the discovery topic and on the HA side to make a JSON value out of a plain string.

What exactly is your proposal? To do the parsing in `e.action()` and still generate a nice JSON payload (in its own topic or as part of `state`)? I'd be OK with that, though I think this should only be the codepath to deal with legacy converters and we should have some new API for the converters to specify the different parts without resorting to string parsing.

**Reply by Koenkk** [reply]
I do not intent do add something additional to the published payloads. My proposal is to remove the regexes out of `homeassistant.ts` and move them into the action exposes (`e.action()`). Also I think we should add new exposes for the other `action_*` values (like [`action_group`](https://github.com/Koenkk/zigbee-herdsman-converters/blob/e7ba34bea6c022ec41acc8e1841effd233ae877a/src/converters/fromZigbee.ts#L4792))

**Reply by mundschenk-at** [reply]
Then to be honest I have no clue what you mean by moving the regexes. What are they supposed to do in `e.action()`? (And I don't see how a separate exposes for `action_group` could be integrated into the HA `event` logic.)

**Reply by mundschenk-at** [reply]
Or do you mean giving each action expose their own parsing regex, the default being something like `^(?:<action>.*)$`? That could work, though I really would prefer having a proper JSON payload for HA instead of my fancy `value_template`. Of course the wildcard actions would have to go.

I still see no way to include `action_group` if it has its own separate expose.

**Reply by Koenkk** [reply]
> Or do you mean giving each action expose their own parsing regex, the default being something like ^(?:<action>.*)$?

Yes that's what I mean, I don't think that having it hardcoded in `homeassistant.ts` is future proof. Also sometimes the group is part of the action, e.g. `{action: "click_1"}`. We could then separate this out with the regex.

> I still see no way to include action_group if it has its own separate expose.

Wouldn't that be as simple as adding `value_json.action_group` to https://github.com/Koenkk/zigbee2mqtt/blob/5b907f2f77490cfcff1a39f5b731d508d0344663/lib/extension/homeassistant.ts#L2271?

**Reply by mundschenk-at** [reply]
But the `action` expose would have to know about the `action_group`, not sure what we would do with a separate expose? It would only be useful if we wanted to have separate HA entity for it. Each expose results in one (or more) entities, the design does not allow for several exposes to be combined into a single discovery topic (HA entity).

**Reply by chris-1243** [reply]
@mundschenk-at Have you found out a solution on how to integrate events for devices such as [511.344](https://github.com/Koenkk/zigbee-herdsman-converters/blob/fcd5cf77df3b0b07c039b840f3faa6c70d0dba04/src/devices/iluminize.ts#L186) ?

**Reply by mundschenk-at** [reply]
@chris-1243 See https://github.com/Koenkk/zigbee2mqtt/discussions/25061#discussioncomment-11566068. But I'm not actively working on anything until an architectural decision has been reached by @Koenkk.

**Reply by chris-1243** [reply]
@mundschenk-at I read the thread already and it completely makes sense not to spend too much effort until a final decision is taken. Thanks for the explanation👍

**Comment by Drafteed** [comment]
To be honest, I would do it like this: сompletely revert #24233 and all followups. Then add a new expose `e.event()` that would correspond to the new HA entity and gradually add them to existing devices/definitions. Also mark `e.action()` as deprecated, removing it in 3.x.x. In this case, the event entity will only be available if it is defined in the converter, otherwise users will have to use MQTT triggers until the event entity is added properly.

Regarding #24233, I think this implementation is a bit messy, IMO. All attempts to make workaround are hacky that will cause a lot of problems that will be difficult to solve later.

**Reply by mundschenk-at** [reply]
Could you be a bit more specific? What elements do you consider "messy" or "hacky"? TBH, I don't see how a new exposes type would make any difference for the implementation if there's not also a new JSON payload (which @Koenkk said he doesn't want to add).

**Reply by Drafteed** [reply]
> Could you be a bit more specific? What elements do you consider "messy" or "hacky"?

I think this YAML could easily pilot a flight to the moon. 😊https://github.com/Koenkk/zigbee2mqtt/blob/7dc92d4b6c40fc1d439aee6a1550daa375929a97/lib/extension/homeassistant.ts#L2271-L2286

Please understand correctly, I do not want to offend you. You have done a big and very great job 👍. However, it seems to me that the current implementation of `action` is not very suitable for creating an `event` entity from it, since it can contain unpredictable data (e.g. wildcards), without any standardization (e.g. `single_right`, `right_l1`, `on_right`, `1_single`, `on_1`, etc.), and any separation (which part of string is the endpoint, and which is the event type?). Every attempt to fix this will cause breaking changes. On the other hand, I suggest creating a new expose that allows for some level of standardization while keeping `action` as is to avoid introducing any breaking changes in the future.

**Reply by mundschenk-at** [reply]
Oh, I agree that the `value_template` currently is potentially a bit fragile (or at least not very elegant), but to fix that, we need a new JSON payload (whether it's a new object in `state` or a separate topic). I don't see how a separate new `exposes` type would help in any way. IMHO it would make the change more onerous because every single converter would need to be changed, and either you lose the MQTT device triggers or you'd have to duplicate the functionality. 

@Koenkk's proposal of moving the parsing regex to the converter/`e.action()` is a good idea, because it can then be hyper-specific to the individual converter where necessary, but allows us to switch over all devices at once (at least on a basic level). To get rid of the `value_template` magic would still need the new JSON payload, though (which I am strongly in favor of!).

**Reply by Drafteed** [reply]
> I don't see how a separate new `exposes` type would help in any way.

Of course I mean that we will have to introduce a brand new property (e.g. `event`) in each converter and the new payload, leaving the `action` as is, to avoid lose the MQTT device triggers.

**Reply by mundschenk-at** [reply]
But you said `action` should be deprecated?

**Reply by Drafteed** [reply]
> But you said `action` should be deprecated?

Exactly.

**Reply by mundschenk-at** [reply]
So that would deprecate the MQTT device triggers. I don't think we should do that.

**Reply by Nerivec** [reply]
Actions cannot be removed, events are HA-only (would need a complete rework of actions otherwise, a huge breaking change for all setups, frontend, node-red, etc.). Hence the annoyance of duplicating code for converters with the introduction of new props... And it shouldn't impact performance for other users (HA extension disabled).
Also, parsing with regexes has to be simple enough, every time a new level of parsing is introduced, that already depends on a regex parsing, you potentially introduce a LOT of edge-cases that tests will undoubtedly fail to detect (i.e. 🐛 abound). Not to mention the difficulty increase for maintenance.

**Reply by Koenkk** [reply]
@Drafteed I guess your `event` proposal is similar to the `action_data` proposal from the OP?

> My initial thought was to add a new action_data property for this (https://github.com/Koenkk/zigbee2mqtt/pull/24233#discussion_r1817842276), but this will add a lot of duplication, so not a good option.

I think this would be too much work and duplication. The `action` should also not be removed, it works well for most of the cases (and also not everyone uses HA). Therefore I think it's the best to add the parsing regexes to the `e.action()` (for those that have more stuff in the action than just a simple `single` for example) and add exposes for e.g. `action_group`. Based on that the Home Assistant extension should be able to discover the events.

**Reply by mundschenk-at** [reply]
@Koenkk I still don't see how adding a separate exposes for `action_group` is supposed to work with the current discovery framework.

**Reply by Koenkk** [reply]
@mundschenk-at whenever we encounter an `action`, we check all the other exposes of that device, in case you find any `action_`, it's added to the discovery payload.

**Reply by mundschenk-at** [reply]
I don't think it's very efficient, but OK, it can be done that way. 

I still don't see the benefit of not adding a new JSON state key (not a separate topic) and handling the parsing entirely on the TS side? Yes, it will duplicate _data_ slightly, but it would remove duplicate _parsing_, which is much more error prone. This would also improve compatibility with consumers that don't have the exact same Jinja2 engine as HA.

**Reply by Koenkk** [reply]
> I still don't see the benefit of not adding a new JSON state key (not a separate topic) and handling the parsing entirely on the TS side?

Could you elaborate a bit more on this?

**Reply by mundschenk-at** [reply]
@Koenkk Sorry, I thought I'd have time to do some work on this over the holidays, but didn't really. Since this has become somewhat urgent with the reactions to the release of 2.0.0, here are my thoughts:

* Home Assistant expects a specific JSON payload for `event` entity "states" (in quotation marks because the events themselves are not technically part of the entity state for the `event` platform). The `action` topic payloads on the other hand are simply strings. So we have to transform the MQTT payload to a JSON object.
* In itself, that is expected by HA's developers, as MQTT sources cannot be expected to already follow HA's specific syntax. HA supports several template functions to work more easily with JSON payloads (e.g. replace a key with another).
* However, since our `action` payload is a legacy flat string, it needs to be transformed to support splitting actual actions and objects (and possibly other attributes). Currently, this is done twice - once on the Z2M side to generate the discovery payload, and once on the HA side for each sent event. To allow that, the JS-Regexes need to converted to Python/Jinja2 syntax and the resulting `value_template` is quite complex ("could easily pilot a flight to the moon", as @Drafteed phrased it). Doing such an expansive transformation in Jinja2 is unfortunately quite brittle:
    - It heavily depends on the template engine. This is less of a an issue for HA itself and more for other consumers of the discovery syntax.
    - I'm not sure how performant the regex transformations are on the HA side (though this point might be reduced if we have more specific regexes directly in the converter).
    - Every tiny change means that the tests will have to be adapted as the `value_template` string changes.
* If we could drop the need to run the regexes on the HA side, the value templates could be much simpler (if needed at all). There would be some duplication of payload data (in that we'd need to create a JSON object in `state` or a separate topic), but in my opinion this would be offset by making maintenance a lot easier.

**Reply by mundschenk-at** [reply]
@Koenkk So something has come up in the [WTH conversation with Petro](https://community.home-assistant.io/t/wth-are-event-entities-not-triggered-a-second-time-when-using-the-attribute-event-type-in-a-state-trigger/804428/27?u=mundschenk-at), which partially makes me reconsider my stance against @Drafteed's proposal: Event entities were intended to be smaller than what we have in Z2M (which is mostly a single `action` per device). It's not clear at all in the [MQTT Event documentation](https://www.home-assistant.io/integrations/event.mqtt/), but the examples in the [Python docs](https://www.home-assistant.io/integrations/event/#event-types) and in the [announcement post](https://www.home-assistant.io/blog/2023/08/02/release-20238/#introducing-the-event-entity) strongly insinuate that each button should have their own entity.

Having a separate entity for each button means that users can target their automations much more precisely using just the UI (though they still have to query the `event_type` attribute with all that entails to limit it to specific presses). The only reasonable way to implement this for Z2M (if we so choose) is to make a very simple `event` entity implementation for the existing `action` expose that just mirrors the `action` topic and start adding a more fine-grained `event` expose (and a corresponding JSON payload in `state`) to the converters one by one. (This new expose type would contain enough information to make each `event` entity to a single button, even if it is a single `expose`, though we could also do multiple exposes if so desired.)

**Reply by Drafteed** [reply]
I still think that we should temporarily remove events from z2m and implement them later - after carefully considering all aspects to avoid problems in the future.
In the current implementation, leave only MQTT triggers and legacy `action` sensor, available by explicitly setting `legacy_action_sensor`.

**Reply by Drafteed** [reply]
> available by explicitly setting `legacy_action_sensor`

BTW, it would be nice to have a warning about the use of deprecated/experemental functions in the log and ideally in the frontend to avoid a surprised reaction from the community when these legacy functions stop working.

**Reply by antst** [reply]
> I still think that we should temporarily remove events from z2m and implement them later - after carefully considering all aspects to avoid problems in the future. In the current implementation, leave only MQTT triggers and legacy `action` sensor, available by explicitly setting `legacy_action_sensor`.

I agree with that. Lets HA to sort out some issues with event entities, then Z2M can support it. To be honest, current legacy action entities is the best ATM. It  is events, how they have to be.

**Reply by D3v01dZA** [reply]
> I agree with that. Lets HA to sort out some issues with event entities, then Z2M can support it.

The way that event entities work in HA and your dislike of it is not a valid reason to remove support. A valid reason might be what they’re alluding to above, which is that it’s creating some extra maintenance burden.

**Reply by antst** [reply]
> > I agree with that. Lets HA to sort out some issues with event entities, then Z2M can support it.
> 
> The way that event entities work in HA and your dislike of it is not a valid reason to remove support. A valid reason might be what they’re alluding to above, which is that it’s creating some extra maintenance burden.

It is not what I am saying. I am saying that from look of it and limitations and discussions around it on HA forums, event entities system in HA is likely to pass through some changes. And better to spent resources on Z2M support of it, when HA will have "v2" of event entities, to avoid redoing stuff in Z2M. Which means to keep current support as experimental and as is.
Otherwise we set event entities as recommended, people start to redo their automation (half year after they redid to device triggers) and then half year later, this has to change again. And you don't want people to redo their automaton every year or half year :)

At the end of the day, it does not matter how complex particular automation looks like(even if I prefer when it looks simple and clean), you make it once and forget. People don't make new automations every day. And once in a while, when they need new automation, they can manage with anything, with help of GH and HA forums. It's only problematic when you need to rewrite dozens of automations suddenly, again and again. 

Personally, I passed already through 3 transitions in how Z2M+HA automations handled. And it is waste of time every time, to rewrite it all. Especially, taking into account there is no added value in that. It is not like suddenly my home will perform space flight after that :) Automation works exactly the same as they used, which is exactly the same as I want. But I need to spend time, to get exactly the same as I had already. Because somebody somewhere decided that there is a better way, which is, often, does not have really motivated technical reasons. But Z2M and HA are production software. Not experimental toys. 

And, you know, every new line of code is a potential new bug :) Avoiding writing those new lines of code without real actuality important reasons is an art of production software :)

**Reply by mundschenk-at** [reply]
I highly doubt that HA will change their `event` platform in any significant way. No "Version 2" of that is coming in the foreseeable future. 

Maybe they'll tweak their state engine for the `event_type` attribute or add a new type of automation trigger for event entities, but at this point I think it's more likely that they'll add some UI sugar instead. @Petro31's [WTH has only 58 votes](https://community.home-assistant.io/t/wth-are-event-entities-not-triggered-a-second-time-when-using-the-attribute-event-type-in-a-state-trigger/804428?u=mundschenk-at) at the moment, so it does not look like there is a lot of momentum for any change at all on the HA side.

**Reply by antst** [reply]
> I highly doubt that HA will change their `event` platform in any significant way. No "Version 2" of that is coming in the foreseeable future.
> 
> Maybe they'll tweak their state engine for the `event_type` attribute or add a new type of automation trigger for event entities, but at this point I think it's more likely that they'll add some UI sugar instead. @Petro31's [WTH has only 58 votes](https://community.home-assistant.io/t/wth-are-event-entities-not-triggered-a-second-time-when-using-the-attribute-event-type-in-a-state-trigger/804428?u=mundschenk-at) at the moment, so it does not look like there is a lot of momentum for any change at all on the HA side.

Their bad! :)

**Comment by antst** [comment]
Just in case, real-life example of evolution of automations in HA with current implementation of event entities. Simple case: button (in this aqara) toggles light group.

legacy actions:
```
triggers:
  - trigger: state
    entity_id: sensor.study_button_action
    enabled: true
    to: single
actions:
  - action: light.toggle
    target:
      entity_id: light.study_lights
```

currently recommended device triggers:

```
triggers:
  - domain: mqtt
    device_id: a6e65827e8df0b0725d86b1384138489
    type: action
    subtype: single
    trigger: device
    enabled: true
actions:
  - action: light.toggle
    target:
      entity_id: light.study_lights
```

event entity:

```
triggers:
  - trigger: state
    entity_id:
      - event.study_button_action
    enabled: true    
conditions:
  - condition: state
    entity_id: event.study_button_action
    attribute: event_type
    state: single
actions:
  - action: light.toggle
    target:
      entity_id: light.study_lights
```

If I add second button to this scenario, this story will look even more scary....

**Reply by D3v01dZA** [reply]
>If I add second button to this scenario, this story will look even more scary....

It seems like you're not a fan of the event style, but is that not a critique of how events work in home assistant in general and not with the current implementation in Z2M?

**Reply by antst** [reply]
> > If I add second button to this scenario, this story will look even more scary....
> 
> It seems like you're not a fan of the event style, but is that not a critique of how events work in home assistant in general and not with the current implementation in Z2M?

hm. thing is, all 3 cases works in HA right now. 

but Z2M seems to be going very rapidly from compact and readable to very uncompact and unreadable. And, to be honest, I miss the point. It does not provide extra functionality. And does not make things simpler. Then, of course, logical question is "why?".

**Reply by D3v01dZA** [reply]
>hm. thing is, all 3 cases works in HA right now.

Yes all 3 cases work, and if it were me I would prefer to use device triggers in HA because of the verbosity of events, but events are a standard way to do things too right now.

>Then, of course, logical question is "why?".

I think events are kind of a standard way to interact with events in HA, ZWaveJS for example implements them too. I'll also give my reason why, I prefer to write a couple of my automations in AppDaemon because it's a lot easier to write re-usable automations there than using blueprints. The event entity there is actually _highly convenient_, monitor one entity and do stuff. For reasons I don't yet understand, apparently device triggers are not available over web sockets so I actually can't use Z2M in the way that I prefer (in AppDaemon via HA) - this is actually a HA problem in that they don't expose triggers but here I am.

**Reply by antst** [reply]
> > hm. thing is, all 3 cases works in HA right now.
> 
> Yes all 3 cases work, and if it were me I would prefer to use device triggers in HA because of the verbosity of events, but events are a standard way to do things too right now.

device triggers are OKish, but problem is they are totally unreadable. Not in yaml, not in UI (unless you expand it all, you can't quickly glance). and totally unwritable manually in yaml files, unless your write some automation for this device in UI, convert to yaml, and get from it device id (at least I did not found the way how to easy get this id).

> > Then, of course, logical question is "why?".
> 
> I think events are kind of a standard way to interact with events in HA, ZWaveJS for example implements them too. I'll also give my reason why, I prefer to write a couple of my automations in AppDaemon because it's a lot easier to write re-usable automations there than using blueprints. The event entity there is actually _highly convenient_, monitor one entity and do stuff. For reasons I don't yet understand, apparently device triggers are not available over web sockets so I actually can't use Z2M in the way that I prefer (in AppDaemon via HA) - this is actually a HA problem in that they don't expose triggers but here I am.

to be honest, events somehow reminds me what was even before first version I have in my post (version we call  legacy for a week). But that implementation didn't require complications for simple cases. Yes, maybe in some cases events (in current form) can make things somehow easier, even being cumbersome, but I doubt it is true in 95% of cases. Looking at situation with events now, if that would be future "default", then, indeed, I can end up writing automations completely outside of scope of HA. But then I'd deal directly with MQTT. But I don't see this is a good case.

**Reply by D3v01dZA** [reply]
> device triggers are OKish, but problem is they are totally unreadable. Not in yaml, not in UI (unless you expand it all, you can't quickly glance). and totally unwritable manually in yaml files, unless your write some automation for this device in UI, convert to yaml, and get from it device id (at least I did not found the way how to easy get this id).

+1. I also really don't like digging around for device ids, I've more than once replaced sensors and been left with a mess because of device ids even though my entity ids are all the same.

> But then I'd deal directly with MQTT. But I don't see this is a good case.

I didn't do this initially thinking I'd like to move back to ZHA if it became necessary but honestly I may just start reading straight from MQTT now that you mention it.

**Reply by antst** [reply]
> I didn't do this initially thinking I'd like to move back to ZHA if it became necessary but honestly I may just start reading straight from MQTT now that you mention it.

But then, again, I can argue that you do it, because dealing with automations in HA (where most of automations supposed to be) is not really comfortable.

And, coming back to event entities. And even triggers. Was there something which could not be implemented in what became now legacy? There should be a reason, to make automations bigger and unreadable. And then, again, there must be a reason to make them even bigger and still unreadable.
Unless, of course, someone trully  believes that making code bigger and more complex makes it easier to manage :)

In reality, event entities do exactly the same as states of legacy action entities. The only practical difference is that you could do all what you have in you example, in the same way, but you could also keeps simple things simple, like in case I cite.

**Reply by mundschenk-at** [reply]
Not to derail your thread maliciously, but this discussion should be about the How, not the Why or If of `event` entities. Most of that verbosity in automation usage would immediately go away if https://community.home-assistant.io/t/wth-are-event-entities-not-triggered-a-second-time-when-using-the-attribute-event-type-in-a-state-trigger/804428 was implemented.

**Reply by antst** [reply]
> Not to derail your thread maliciously, but this discussion should be about the How, not the Why or If of `event` entities. Most of that verbosity in automation usage would immediately go away if https://community.home-assistant.io/t/wth-are-event-entities-not-triggered-a-second-time-when-using-the-attribute-event-type-in-a-state-trigger/804428 was implemented.

Yes, agree with that. That would, effectively restore combination of new entities and HA to work exactly as legacy action entities works :) the only difference double change of "sensor.xxx_action" to "event.xxx_action".

**Reply by D3v01dZA** [reply]
>Not to derail your thread maliciously

I don't think you did, but it does seem like deleting the code for events would technically be a 'way forward' so addressing the 'why' does seem to have some place here. I'm in the position that I'd prefer to see events stay around which is why I threw my hat in the ring.

**Reply by mundschenk-at** [reply]
That was not under discussion, as `event` entities are the HA recommended solution for handling momentary actions.

Addendum: A more [detailed reasoning](https://github.com/Koenkk/zigbee2mqtt/pull/24233) can be found in the original PR.

**Reply by antst** [reply]
So, outcome is simple. 

1) Either HA should change events implementation, when it is allows to react to from X to X, even if X is unchanged. But that's goes against logic they seems to trying implement, where action happens on state change.

2) Or Z2M should send "cancellation" message (empty string or something predefined), but that will be a hell, because for every device logic will be different. For hue, for example, one does not need to do something, and for Aqara button, it has to be done.
Moreover, there are some devices, if I recall right, which sends both, press-release, and click. Then it is even unclear, how to handle :) this will overcomplicate converters without real reason.

But before one of those happens, it is "dead end for masses".

But, for the gods sake, I miss the point why HA team decided to bring additional complication. Essentially, it is no different from state change trigger. With the exception that forcing it to work only in state change (which is totally possible to have with normal  state change automation).

**Reply by D3v01dZA** [reply]
> Either HA should change events implementation, when it is allows to react to from X to X, even if X is unchanged.

Events already allow this, they are triggered by the state which I implemented as a time much like a button and have the event as an attribute.

> Or Z2M should send "cancellation" message.

Not necessary because of what above.

**Reply by Koenkk** [reply]
@jbouwh just confirmed that event entities should trigger even when receiving the same event twice. @mundschenk-at could we be doing something wrong from the z2m side?

**Reply by mundschenk-at** [reply]
I think this might be a misunderstanding of the issue: The `event` entities trigger fine with the same event twice in a row, however that just means the _state_ (in HA parlance, not Z2M's) of that entity gets updated to the current timestamp. The _attribute_ `event_type` of that entity stays unchanged (because it's the same event and HA silently drops unchanged attribute values), and anything listening to an _attribute_ change will not get triggered.

Technically, that is all consistent, but somewhat unexpected behavior for at least for non-technical users. It also means to get all events from the entity, you have base your automation trigger on the `event` entity state (i.e. all events) and need a condition afterwards to narrow things to a specific `event_type`. This is not very elegant or concise, especially considering that it needs templating to do it properly without race conditions (i.e. you cannot do it all in the UI).

@Koenkk @jbouwh

**Reply by DubhAd** [reply]
There _is_ discussion about this "feature" - <https://community.home-assistant.io/t/wth-are-event-entities-not-triggered-a-second-time-when-using-the-attribute-event-type-in-a-state-trigger/804428/>. The OP there has started trying to get the behaviour made sane, at least as an option people can enable, but for now I'm afraid it's really best done with templates:
```yaml
triggers:
  - platform: state
    entity_id: sensor.office_dial_moes_action
    to: 'toggle'
```
becomes
```yaml
triggers:
  - trigger: state
    entity_id: event.office_dial_moes_action
    to: ~
        # null and ~ (and NULL) are functionally identical - I just happen to prefer ~
conditions:
  - condition: template
    value_template: >-
      {{ trigger.to_state.attributes.event_type == 'toggle' }}
```

The reason for using templates is so that you're looking at the state _that caused the trigger to fire_, otherwise if multiple things happen you could be looking at the wrong state.

That can at least all be done in the UI (I wrote it up [here](https://blog.ceard.tech/2025/01/zigbee2mqtt-2.0)).

**Reply by mundschenk-at** [reply]
Only if you consider templating to be "in the UI" 🤷

**Reply by jbouwh** [reply]
I get that it is some complex to configure triggers based on event entities. That might be something to improve in HA, so that is more easy to just act on an event entity being triggered.

**Reply by jbouwh** [reply]
I mean something like: https://www.home-assistant.io/docs/configuration/events/
For the HA IMAP integration I use the same technique. Improvent starts with a discussion, So may be I just should make a proposal to it? It might even lead to a new trigger type eventually.

**Reply by mundschenk-at** [reply]
Hi @jbouwh! What do you mean to propose? I'm not sure what you are referring to with that link to HA events (note for others in this thread: not `event` entities).

**Reply by Petro31** [reply]
> Hi @jbouwh! What do you mean to propose? I'm not sure what you are referring to with that link to HA events (note for others in this thread: not `event` entities).

He's talking about making a proposal to the core HA team about event entities.

**Reply by mundschenk-at** [reply]
> I mean something like: https://www.home-assistant.io/docs/configuration/events/
> For the HA IMAP integration I use the same technique.

Yes, of course, but what?

**Reply by Petro31** [reply]
> > I mean something like: https://www.home-assistant.io/docs/configuration/events/
> > For the HA IMAP integration I use the same technique.
> 
> Yes, of course, but what?

Based on what he linked, he may be referring to "old school events".  

To be quite honest, I prefer them.  However they are hard for beginners to understand, which is why event entities were created in the first place.

**Reply by jbouwh** [reply]
Right, but some how event entities need an easier way to configure a trigger with them. The old events have some advantages that they can propagate data attributes from the event, that can be used to filter on.

**Reply by mundschenk-at** [reply]
@jbouwh Any news?

@Koenkk I think we should decide on a way forward, whatever it is. The current PoC implementation should not stay as a permanent temporary solution.

**Reply by jbouwh** [reply]
> @jbouwh Any news?

Nope. I think base is okay, but the trigger it self should have is own type. Using the state is not enough, and just testing on the attributes does not work either.

**Reply by jbouwh** [reply]
Started an architectural discussion here: https://github.com/home-assistant/architecture/discussions/1225

**Reply by mundschenk-at** [reply]
And it's already closed. Oh well.

**Reply by jbouwh** [reply]
Yeah, but there's some light in the tunnel, as the trigger system is being improved.

**Comment by thk-socal** [comment]
Started playing with the event entities. While I see both sides, one thing that was amazing for me was the ability to teach my wife how to make her own automations in the HA GUI tied to events of my Zigbee switches. I showed her how to press the button to force HA to see the event as an option, then go to the Automation GUI and choose that switch by name and pick the event to then drive any automation. I realize the YAML way is harder, but the GUI option was amazing for her and made it easy for her to make her own things. 

So, that helps drive even more acceptance of the automations to my home. More people can more easily do it, which I think is a good thing. Currently have 2 Zigbee networks, each exceeding 180 devices and Z2M is the best for me, and I love the direction with events.

**Comment by mundschenk-at** [comment]
@Koenkk We (meaning: you) really should decide what to do. I believe all arguments have been exchanged on the general level. There are lots of inconsistencies with the current implementation, but I don't want to work on fixing them when the underlying direction has not been chosen. 

I don't think waiting for Home Assistant Core regarding a possible trigger rework is helpful on the Z2M side. (They won't change their basic architecture of events even if something moves forward there.)

**Reply by Koenkk** [reply]
I still think the regex approach is the best, but let's discuss offline, can you contact me on discord or telegram? (@koenkk)

**Reply by mundschenk-at** [reply]
To document the follow-up: We did have a discussion, but I never got around to write the PR. In the meantime, [purpose-specific triggers and conditions](https://www.home-assistant.io/blog/2026/07/01/release-20267/#creating-automations-just-became-easier-natural-and-more-powerful) have made some of the objections regarding ease-of-use of `event` entities obsolete.
